docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+Custom Metrics API
+==================
+
+The new [metrics monitoring vision]() proposes an API that the Horizontal
+Pod Autoscaler can use to access arbitrary metrics.
+
+Similarly to the [master metrics API](resource-metrics-api.md), the new
+API should be structured around accessing metrics by referring to
+kubernetes objects (or groups thereof) and a metric name.  For this
+reason, the API could be useful for other consumers (most likely
+controllers) that want to consume custom metrics (similarly to how the
+master metrics API is generally useful to multiple cluster components).
+
+The HPA can currently refer to all pods selected by a label selector, and
+is planned to be able to support referring to metrics describing
+a namespace and other non-pod objects within the namespace.
+
+API Paths
+---------
+
+The root API path will look like `/apis/general-metrics/v1alpha1`.  For
+brevity, this will be left off below.
+
+- `/nodes/{node-name}/metrics/{metric-name...}`: retrieve the given metric
+  on the given node
+
+- `/namespaces/{namespace-name}/metrics/{metric-name...}`: retrieve the
+  given metric on the given namespace
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}`:
```

Comment:
I should probably clarify what `object-type` means in a multi-API-group world -- I was thinking that it would be the string form of the `GroupKind` object in Kubernetes.  This does potentially introduce the requirement to recognize objects moved between API groups, but I don't think there's any good way around that.